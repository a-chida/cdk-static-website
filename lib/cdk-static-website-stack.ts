import * as cdk from "aws-cdk-lib";
import * as cloudfront from "aws-cdk-lib/aws-cloudfront";
import * as origins from "aws-cdk-lib/aws-cloudfront-origins";
import * as iam from "aws-cdk-lib/aws-iam";
import * as s3 from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export class CdkStaticWebsiteStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // S3 bucket
    const s3Bucket = new s3.Bucket(this, "S3Bucket", {
      autoDeleteObjects: true,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      bucketName: "cdk-static-website-origin",
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // CloudFront
    const origin = new origins.S3Origin(s3Bucket);
    const distribution = new cloudfront.Distribution(this, "distribution", {
      defaultBehavior: { origin: origin },
      defaultRootObject: "index.html",
      geoRestriction: cloudfront.GeoRestriction.allowlist("JP"),
      httpVersion: cloudfront.HttpVersion.HTTP2_AND_3,
      minimumProtocolVersion: cloudfront.SecurityPolicyProtocol.TLS_V1_2_2021,
      priceClass: cloudfront.PriceClass.PRICE_CLASS_200,
    });

    // Use L1 constructor to implement OAC because L2 constructor is not support Origin Access Control
    // Ref: https://github.com/aws/aws-cdk/issues/21771
    const cfnOriginAccessControl = new cloudfront.CfnOriginAccessControl(
      this,
      "cf-origin-access-control",
      {
        originAccessControlConfig: {
          name: s3Bucket.bucketRegionalDomainName,
          originAccessControlOriginType: "s3",
          signingBehavior: "always",
          signingProtocol: "sigv4",
          description: "S3 Access Control",
        },
      }
    );

    // Additional settings for origin 0 (0: s3Bucket)
    const cfnDistribution = distribution.node
      .defaultChild as cloudfront.CfnDistribution;
    // Delete OAI
    cfnDistribution.addOverride(
      "Properties.DistributionConfig.Origins.0.S3OriginConfig.OriginAccessIdentity",
      ""
    );
    // OAC does not require CustomOriginConfig
    cfnDistribution.addPropertyDeletionOverride(
      "DistributionConfig.Origins.0.CustomOriginConfig"
    );
    // By default, the s3 WebsiteURL is set and an error occurs, so set the S3 domain name
    cfnDistribution.addPropertyOverride(
      "DistributionConfig.Origins.0.DomainName",
      s3Bucket.bucketRegionalDomainName
    );

    // OAC settings
    cfnDistribution.addOverride(
      "DistributionConfig.Origins.0.OriginAccessControlId",
      cfnOriginAccessControl.getAtt("Id")
    );

    // add S3 bucket policy for CloudFront
    s3Bucket.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: "AllowCloudFrontServicePrincipalReadOnly",
        effect: iam.Effect.ALLOW,
        principals: [new iam.ServicePrincipal("cloudfront.amazonaws.com")],
        actions: ["s3:GetObject"],
        resources: [s3Bucket.arnForObjects("*")],
        conditions: {
          StringEquals: {
            "aws:SourceArn": `arn:aws:cloudfront::${this.account}:distribution/${distribution.distributionId}`,
          },
        },
      })
    );
  }
}
