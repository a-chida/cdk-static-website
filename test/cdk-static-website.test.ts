import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as CdkStaticWebsite from "../lib/cdk-static-website-stack";

describe("CdkStaticWebsite", () => {
  const app = new cdk.App();
  const stack = new CdkStaticWebsite.CdkStaticWebsiteStack(app, "TestStack", {
    env: { region: "us-east-1" },
  });
  const template = Template.fromStack(stack);

  // snapshot test
  test("snapshot test", () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

  // S3 Bucket
  test("S3 Bucket", () => {
    template.hasResourceProperties("AWS::S3::Bucket", {
      PublicAccessBlockConfiguration: {
        BlockPublicAcls: true,
        BlockPublicPolicy: true,
        IgnorePublicAcls: true,
        RestrictPublicBuckets: true,
      },
    });
  });

  // CloudFront Distribution
  test("CloudFront Distribution", () => {
    template.findResources("AWS::CloudFront::Distribution");
  });

  // CloudFront OriginAccessControl
  test("CloudFront OriginAccessControl", () => {
    template.hasResourceProperties("AWS::CloudFront::OriginAccessControl", {
      OriginAccessControlConfig: {
        OriginAccessControlOriginType: "s3",
      },
    });
  });
});
