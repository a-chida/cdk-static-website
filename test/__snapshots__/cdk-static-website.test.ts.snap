// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdkStaticWebsite snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-us-east-1",
          },
          "S3Key": "2eb6a831b107939f63cfebf68e6316e1a40f79fc99cae0fee9b333bac8d29bc4.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "S3Bucket07682993",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Function76856677": {
      "Properties": {
        "AutoPublish": true,
        "FunctionCode": "async function handler(event) {
  const request = event.request;
  const uri = request.uri;

  // Check whether the URI is missing a file name.
  if (uri.endsWith("/")) {
    request.uri += "index.html";
  }
  // Check whether the URI is missing a file extension.
  else if (!uri.includes(".")) {
    request.uri += "/index.html";
  }

  return request;
}
",
        "FunctionConfig": {
          "Comment": "cdk-static-website-url-rewrite-spa",
          "Runtime": "cloudfront-js-1.0",
        },
        "Name": "cdk-static-website-url-rewrite-spa",
      },
      "Type": "AWS::CloudFront::Function",
    },
    "S3Bucket07682993": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketName": "cdk-static-website-origin",
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "S3BucketAutoDeleteObjectsCustomResource7735AB63": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "S3BucketPolicyF560589A",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "S3Bucket07682993",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "S3BucketPolicyF560589A": {
      "Properties": {
        "Bucket": {
          "Ref": "S3Bucket07682993",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "S3Bucket07682993",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "S3Bucket07682993",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "distributionOrigin1S3OriginC4E11128",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "S3Bucket07682993",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            {
              "Action": "s3:GetObject",
              "Condition": {
                "StringEquals": {
                  "aws:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:cloudfront::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":distribution/",
                        {
                          "Ref": "distribution114A0A2A",
                        },
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudfront.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "S3Bucket07682993",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
              "Sid": "AllowCloudFrontServicePrincipalReadOnly",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "cforiginaccesscontrol": {
      "Properties": {
        "OriginAccessControlConfig": {
          "Description": "S3 Access Control",
          "Name": {
            "Fn::GetAtt": [
              "S3Bucket07682993",
              "RegionalDomainName",
            ],
          },
          "OriginAccessControlOriginType": "s3",
          "SigningBehavior": "always",
          "SigningProtocol": "sigv4",
        },
      },
      "Type": "AWS::CloudFront::OriginAccessControl",
    },
    "distribution114A0A2A": {
      "DistributionConfig": {
        "Origins": {
          "0": {
            "OriginAccessControlId": {
              "Fn::GetAtt": [
                "cforiginaccesscontrol",
                "Id",
              ],
            },
          },
        },
      },
      "Properties": {
        "DistributionConfig": {
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "FunctionAssociations": [
              {
                "EventType": "viewer-request",
                "FunctionARN": {
                  "Fn::GetAtt": [
                    "Function76856677",
                    "FunctionARN",
                  ],
                },
              },
            ],
            "TargetOriginId": "TestStackdistributionOrigin1B8BE8DD3",
            "ViewerProtocolPolicy": "allow-all",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2and3",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "S3Bucket07682993",
                  "RegionalDomainName",
                ],
              },
              "Id": "TestStackdistributionOrigin1B8BE8DD3",
              "S3OriginConfig": {
                "OriginAccessIdentity": "",
              },
            },
          ],
          "PriceClass": "PriceClass_200",
          "Restrictions": {
            "GeoRestriction": {
              "Locations": [
                "JP",
              ],
              "RestrictionType": "whitelist",
            },
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "distributionOrigin1S3OriginC4E11128": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Identity for TestStackdistributionOrigin1B8BE8DD3",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
